File: BookServiceJava/src/main/java/org/elcata98/mybooks/booksservice/response/BookResponseEntityBuilder.java
package org.elcata98.mybooks.booksservice.response;

import org.elcata98.mybooks.booksservice.model.Book;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;

import javax.servlet.http.HttpServletRequest;
import java.net.URI;
import java.net.URISyntaxException;


@Component
public class BookResponseEntityBuilder implements ResponseEntityBuilder<Book> {

    @Override
    public ResponseEntity<Response<Book>> buildCreateResponseEntity(final Book instance, final HttpServletRequest httpServletRequest) throws URISyntaxException {

        return ResponseEntity.created(new URI(httpServletRequest.getRequestURI() + "/" + instance.getId())).body(new Response<>(instance));
    }

    @Override
    public ResponseEntity<Response<Book>> buildUpdateResponseEntity(final Book instance) {

        return buildInstanceBasedResponseEntity(instance);
    }

    @Override
    public ResponseEntity<Response<Book>> buildDeleteResponseEntity(final boolean success) {

        if (success) {
            return ResponseEntity.noContent().build();
        } else {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    @Override
    public ResponseEntity<Response<Book>> buildGetResponseEntity(final Book instance) {

        return buildInstanceBasedResponseEntity(instance);
    }


    private ResponseEntity<Response<Book>> buildInstanceBasedResponseEntity(final Book instance) {

        if (instance == null) {
            return ResponseEntity.notFound().build();
        } else {
            return ResponseEntity.ok(new Response<>(instance));
        }
    }
}

File: BooksService/src/main/kotlin/org/elcata98/mybooks/booksservice/response/BookResponseEntityBuilder.kt
package org.elcata98.mybooks.booksservice.response

import org.elcata98.mybooks.booksservice.model.Book
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Component
import java.net.URI
import javax.servlet.http.HttpServletRequest


@Component
class BookResponseEntityBuilder : ResponseEntityBuilder<Book> {

    override fun buildCreateResponseEntity(instance: Book, httpServletRequest: HttpServletRequest): ResponseEntity<Response<Book>> {

        return ResponseEntity.created(URI(httpServletRequest.requestURI + "/" + instance.bookId)).body(Response(instance))
    }

    override fun buildUpdateResponseEntity(instance: Book?): ResponseEntity<Response<Book>> {

        return buildInstanceBasedResponseEntity(instance)
    }

    override fun buildDeleteResponseEntity(success: Boolean): ResponseEntity<Response<Book>> {

        return if (success) {
            ResponseEntity.noContent().build()
        } else {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build()
        }
    }

    override fun buildGetResponseEntity(instance: Book?): ResponseEntity<Response<Book>> {

        return buildInstanceBasedResponseEntity(instance)
    }

    private fun buildInstanceBasedResponseEntity(instance: Book?): ResponseEntity<Response<Book>> {

        return if (instance == null) {
            ResponseEntity.notFound().build()
        } else {
            ResponseEntity.ok(Response(instance))
        }
    }
}